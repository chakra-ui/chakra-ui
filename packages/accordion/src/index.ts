export * from "./accordion"
export * from "./use-accordion"
