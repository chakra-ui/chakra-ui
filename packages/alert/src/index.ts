export { Alert, type AlertProps } from "./alert"
export { useAlertStyles, type AlertStatus } from "./alert-context"
export {
  AlertDescription,
  type AlertDescriptionProps,
} from "./alert-description"
export { AlertIcon, type AlertIconProps } from "./alert-icon"
export { AlertTitle, type AlertTitleProps } from "./alert-title"
