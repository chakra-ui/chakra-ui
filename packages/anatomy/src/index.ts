import { anatomy } from "@chakra-ui/theme-tools"

/**
 * **Accordion anatomy**
 * - Root: the root container of the accordion
 * - Container: the accordion item contains the button and panel
 * - Button: the button is the trigger for the panel
 * - Panel: the panel is the content of the accordion item
 * - Icon: the expanded/collapsed icon
 */
export const accordionAnatomy = anatomy("accordion")
  .parts("root", "container", "button", "panel")
  .extend("icon")

/**
 * **Alert anatomy**
 * - Title: the alert's title
 * - Description: the alert's description
 * - Icon: the alert's icon
 */
export const alertAnatomy = anatomy("alert")
  .parts("title", "description", "container")
  .extend("icon")

/**
 * **Avatar anatomy**
 * - Container: the container for the avatar
 * - Label: the avatar initials text
 * - Excess Label: the label or text that represents excess avatar count.
 * Typically used in avatar groups.
 * - Group: the container for the avatar group
 */
export const avatarAnatomy = anatomy("avatar")
  .parts("label", "badge", "container")
  .extend("excessLabel", "group")

/**
 * **Breadcrumb anatomy**
 * - Item: the container for a breadcrumb item
 * - Link: the element that represents the breadcrumb link
 * - Container: the container for the breadcrumb items
 * - Separator: the separator between breadcrumb items
 */
export const breadcrumbAnatomy = anatomy("breadcrumb")
  .parts("link", "item", "container")
  .extend("separator")

export const buttonAnatomy = anatomy("button").parts()

export const checkboxAnatomy = anatomy("checkbox")
  .parts("control", "icon", "container")
  .extend("label")

export const circularProgressAnatomy = anatomy("progress")
  .parts("track", "filledTrack")
  .extend("label")

export const drawerAnatomy = anatomy("drawer")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const editableAnatomy = anatomy("editable").parts(
  "preview",
  "input",
  "textarea",
)

export const formAnatomy = anatomy("form").parts(
  "container",
  "requiredIndicator",
  "helperText",
)

export const formErrorAnatomy = anatomy("formError").parts("text", "icon")

export const inputAnatomy = anatomy("input").parts("addon", "field", "element")

export const listAnatomy = anatomy("list").parts("container", "item", "icon")

export const menuAnatomy = anatomy("menu")
  .parts("button", "list", "item")
  .extend("groupTitle", "command", "divider")

export const modalAnatomy = anatomy("modal")
  .parts("overlay", "dialogContainer", "dialog")
  .extend("header", "closeButton", "body", "footer")

export const numberInputAnatomy = anatomy("numberinput").parts(
  "root",
  "field",
  "stepperGroup",
  "stepper",
)

export const pinInputAnatomy = anatomy("pininput").parts("field")

export const popoverAnatomy = anatomy("popover")
  .parts("content", "header", "body", "footer")
  .extend("popper", "arrow", "closeButton")

export const progressAnatomy = anatomy("progress").parts(
  "label",
  "filledTrack",
  "track",
)

export const radioAnatomy = anatomy("radio").parts(
  "container",
  "control",
  "label",
)

export const selectAnatomy = anatomy("select").parts("field", "icon")

export const sliderAnatomy = anatomy("slider").parts(
  "container",
  "track",
  "thumb",
  "filledTrack",
)

export const statAnatomy = anatomy("stat").parts(
  "container",
  "label",
  "helpText",
  "number",
  "icon",
)

export const switchAnatomy = anatomy("switch").parts(
  "container",
  "track",
  "thumb",
)

export const tableAnatomy = anatomy("table").parts(
  "table",
  "thead",
  "tbody",
  "tr",
  "th",
  "td",
  "tfoot",
  "caption",
)

export const tabsAnatomy = anatomy("tabs").parts(
  "root",
  "tab",
  "tablist",
  "tabpanel",
  "tabpanels",
  "indicator",
)

/**
 * **Tag anatomy**
 * - Container: the container for the tag
 * - Label: the text content of the tag
 * - closeButton: the close button for the tag
 */
export const tagAnatomy = anatomy("tag").parts(
  "container",
  "label",
  "closeButton",
)
