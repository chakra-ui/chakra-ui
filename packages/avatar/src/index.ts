export * from "./avatar"
export * from "./avatar-group"
