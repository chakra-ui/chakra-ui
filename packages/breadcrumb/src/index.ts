export { Breadcrumb, type BreadcrumbProps } from "./breadcrumb"
export { BreadcrumbItem, type BreadcrumbItemProps } from "./breadcrumb-item"
export {
  BreadcrumbSeparator,
  type BreadcrumbSeparatorProps,
} from "./breadcrumb-separator"
export { BreadcrumbLink, type BreadcrumbLinkProps } from "./breadcrumb-link"
export { useBreadcrumbStyles } from "./breadcrumb-context"
export type { BreadcrumbOptions } from "./breadcrumb-types"
