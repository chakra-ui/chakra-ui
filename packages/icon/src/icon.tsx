import {
  chakra,
  ChakraProps,
  forwardRef,
  SystemStyleObject,
} from "@chakra-ui/system"
import { cx, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

const fallbackIcon = {
  path: (
    <g stroke="currentColor" strokeWidth="1.5">
      <path
        strokeLinecap="round"
        fill="none"
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
      />
      <path
        fill="currentColor"
        strokeLinecap="round"
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
      />
      <circle fill="none" strokeMiterlimit="10" cx="12" cy="12" r="11.25" />
    </g>
  ),
  viewBox: "0 0 24 24",
}

export interface IconProps
  extends Omit<React.SVGAttributes<SVGElement>, keyof ChakraProps>,
    ChakraProps {}

export const Icon = forwardRef<IconProps, "svg">((props, ref) => {
  const {
    as: element,
    viewBox,
    color = "currentColor",
    focusable = false,
    children,
    className,
    __css,
    ...rest
  } = props

  const _className = cx("chakra-icon", className)

  const styles: SystemStyleObject = {
    w: "1em",
    h: "1em",
    display: "inline-block",
    lineHeight: "1em",
    flexShrink: 0,
    color,
    ...__css,
  }

  const shared: any = {
    ref,
    focusable,
    className: _className,
    __css: styles,
  }

  const _viewBox = viewBox ?? fallbackIcon.viewBox

  /**
   * If you're using an icon library like `react-icons`.
   * Note: anyone passing the `as` prop, should manage the `viewBox` from the external component
   */
  if (element && typeof element !== "string") {
    return <chakra.svg as={element} {...shared} {...rest} />
  }

  const _path = (children ?? fallbackIcon.path) as React.ReactNode

  return (
    <chakra.svg verticalAlign="middle" viewBox={_viewBox} {...shared} {...rest}>
      {_path}
    </chakra.svg>
  )
})

if (__DEV__) {
  Icon.displayName = "Icon"
}

export default Icon
