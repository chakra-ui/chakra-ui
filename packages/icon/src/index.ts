export * from "./icon"
export { default } from "./icon"
export * from "./create-icon"
