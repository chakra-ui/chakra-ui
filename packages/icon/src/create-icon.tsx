import { forwardRef } from "@chakra-ui/system"
import { __DEV__ } from "@chakra-ui/utils"
import * as React from "react"
import { Icon, IconProps } from "./icon"

interface CreateIconOptions {
  /**
   * The icon `svg` viewBox
   * @default "0 0 24 24"
   */
  viewBox?: string
  /**
   * The `svg` path or group element
   * @type React.ReactElement | React.ReactElement[]
   */
  path?: React.ReactElement | React.ReactElement[]
  /**
   * If the has a single path, simply copy the path's `d` attribute
   */
  d?: string
  /**
   * The display name useful in the dev tools
   */
  displayName?: string
  /**
   * Default props automatically passed to the component; overwriteable
   */
  defaultProps?: IconProps
}

export function createIcon(options: CreateIconOptions) {
  const {
    viewBox = "0 0 24 24",
    d: pathDefinition,
    path,
    displayName,
    defaultProps = {},
  } = options

  const Comp = forwardRef<IconProps, "svg">((props, ref) => (
    <Icon ref={ref} viewBox={viewBox} {...defaultProps} {...props}>
      {path ?? <path fill="currentColor" d={pathDefinition} />}
    </Icon>
  ))

  if (__DEV__) {
    Comp.displayName = displayName
  }

  return Comp
}
