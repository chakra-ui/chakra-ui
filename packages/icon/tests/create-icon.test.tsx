import * as React from "react"
import { render } from "@chakra-ui/test-utils"
import { createIcon } from "../src"

const chakraIconPath = [
  <title key="1">Chakra Logo</title>,
  <rect
    width="257"
    height="257"
    fill="url(#paint0_linear)"
    rx="128.5"
    key="2"
  />,
  <path
    fill="#fff"
    d="M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"
    key="3"
  />,
  <defs key="4">
    <linearGradient
      id="paint0_linear"
      x1="128.5"
      x2="128.5"
      y2="257"
      gradientUnits="userSpaceOnUse"
    >
      <stop offset="0" stopColor="#7BCBD4" />
      <stop offset="1" stopColor="#29C6B7" />
    </linearGradient>
  </defs>,
]

it("sets default viewBox", () => {
  const Icon = createIcon({
    path: chakraIconPath,
  })

  const { container } = render(<Icon />)

  expect(container.querySelector("svg")).toHaveAttribute("viewBox")
})

it("forwards d attribute onto a path element", () => {
  const d =
    "M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"

  const Icon = createIcon({
    d,
  })

  const { container } = render(<Icon />)

  expect(container.querySelector("path")).toHaveAttribute("d", d)
})

it("accepts a single path", () => {
  const d =
    "M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"
  const fill = "#fff"

  const Icon = createIcon({
    path: <path fill={fill} d={d} />,
  })

  const { container } = render(<Icon />)

  expect(container.querySelector("svg")!.children).toHaveLength(1)
  expect(container.querySelector("path")).toHaveAttribute("d", d)
  expect(container.querySelector("path")).toHaveAttribute("fill", fill)
})

it("accepts multiple paths", () => {
  const Icon = createIcon({
    path: chakraIconPath,
  })

  const { container } = render(<Icon />)

  expect(container.querySelector("svg")!.children).toHaveLength(
    chakraIconPath.length,
  )
})

it("forwards displayName", () => {
  const displayName = "dummy-display-name"

  const Icon = createIcon({
    path: chakraIconPath,
    displayName,
  })

  expect(Icon.displayName).toBe(displayName)
})
