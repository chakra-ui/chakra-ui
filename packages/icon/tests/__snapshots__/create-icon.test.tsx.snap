// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts a single path 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  vertical-align: middle;
}

<svg
    class="chakra-icon emotion-0"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"
      fill="#fff"
    />
  </svg>
</DocumentFragment>
`;

exports[`accepts multiple paths 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  vertical-align: middle;
}

<svg
    class="chakra-icon emotion-0"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <title>
      Chakra Logo
    </title>
    <rect
      fill="url(#paint0_linear)"
      height="257"
      rx="128.5"
      width="257"
    />
    <path
      d="M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"
      fill="#fff"
    />
    <defs>
      <lineargradient
        gradientUnits="userSpaceOnUse"
        id="paint0_linear"
        x1="128.5"
        x2="128.5"
        y2="257"
      >
        <stop
          offset="0"
          stop-color="#7BCBD4"
        />
        <stop
          offset="1"
          stop-color="#29C6B7"
        />
      </lineargradient>
    </defs>
  </svg>
</DocumentFragment>
`;

exports[`forwards d attribute onto a path element 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  vertical-align: middle;
}

<svg
    class="chakra-icon emotion-0"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"
      fill="currentColor"
    />
  </svg>
</DocumentFragment>
`;

exports[`forwards default props 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  vertical-align: middle;
  fill: none;
}

<svg
    class="chakra-icon emotion-0"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <title>
      Chakra Logo
    </title>
    <rect
      fill="url(#paint0_linear)"
      height="257"
      rx="128.5"
      width="257"
    />
    <path
      d="M69.56 133.99l87.59-87c1.64-1.62 4.27.36 3.17 2.38l-32.6 59.76a2 2 0 001.75 2.95h56.34a2 2 0 011.36 3.47l-98.72 92.14c-1.78 1.65-4.41-.68-2.99-2.64l46.74-64.47a2 2 0 00-1.62-3.18H70.97a2 2 0 01-1.41-3.41z"
      fill="#fff"
    />
    <defs>
      <lineargradient
        gradientUnits="userSpaceOnUse"
        id="paint0_linear"
        x1="128.5"
        x2="128.5"
        y2="257"
      >
        <stop
          offset="0"
          stop-color="#7BCBD4"
        />
        <stop
          offset="1"
          stop-color="#29C6B7"
        />
      </lineargradient>
    </defs>
  </svg>
</DocumentFragment>
`;
